Модель управления складом

mikhailmineev opened the pull request.

Commit:
Добавил модель в первом приближении (aurusov/raox-models#16)

Commit:
Добавил модель в первом приближении (aurusov/raox-models#16)

Commit:
Добавил наброски взаимодействия с базой (aurusov/raox-models#16)

Commit:
Merge branch 'master' of https://github.com/mikhailmineev/raox-models

Commit:
Фиск после неудачного merge (aurusov/raox-models#16)

Commit:
EntityManager запускается (aurusov/raox-models#16)

Commit:
Данные успешно читаются из базы (aurusov/raox-models#16)

aurusov commented on warehouse/.classpath:
```diff
@@ -0,0 +1,10 @@
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.xtend.XTEND_CONTAINER"/>
+	<classpathentry kind="lib" path="C:/Users/Михаил/src/raox/ru.bmstu.rk9.rao.lib/bin/" sourcepath="C:/Users/Михаил/src/raox/ru.bmstu.rk9.rao.lib"/>
+	<classpathentry kind="lib" path="C:/Users/Михаил/src/raox/ru.bmstu.rk9.rao.lib/thirdparty/querydsl/lib/querydsl-jpa-4.1.3-apt-hibernate-one-jar.jar"/>
```
Хотелось бы относительный путей. Попробуй относительно корня Эклипса, как [тут](https://github.com/aurusov/raox-models/blob/master/barber/barber_activity/.classpath#L6).

mikhailmineev replied:
https://github.com/mikhailmineev/raox/blob/ba3b8528c8d330fd04d62ca35f1737ce5663510c/ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/execution/BuildUtil.java#L173-L181
Тут представлено два режима генерации данных ссылок, PRODUCTION и DEVELOPMENT, на данный момент у меня вызывается DEVELOPMENT при rebuild, смысл править это появится только после релиза фичи с чтением из базы

mikhailmineev replied:
Еще вопрос насчет релиза, все библиотеки thirdparty будут залиты в ru.bmstu.rk9.rao.lib.jar? тогда в релизе не понадобится прописывать querydsl

aurusov replied:
>смысл править это появится только после релиза фичи с чтением из базы

Править для релиза придется не java-код, а условный инсталлятор. Чтобы нудная либа лежала в `ECLIPSE_HOME/dropins/`.

aurusov replied:
>Еще вопрос насчет релиза, все библиотеки thirdparty будут залиты в ru.bmstu.rk9.rao.lib.jar? тогда в релизе не понадобится прописывать querydsl

Как вариант. Посмотри использование `commons-math3-3.6.jar` через поиск по всем файлам по содержимому.

aurusov commented on warehouse/.project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>warehouse</name>
```
Название уж больно общее. Вторая модель склада сразу будет конфликтовать.

mikhailmineev replied:
Тогда грамотнее будет сделать подпапки warehouse->warehouse-model ну и warehouse->warehouse-receipt (заказ)?

aurusov replied:
Общая пупка нужна, но не факт, что `warehouse`. Надо Андрея спросить. Вернемся к этому месту чуть позже.

aurusov replied:
@Troideninc

Подскажи название модели. Хочется, чтобы оно отражало именно твою систему. М.б. прям юридическое название фирмы ?

Troideninc replied:
Предлагаю: Русское название "Модель поставки под заказ"
Английское "delivery to order"

aurusov replied:
@Troideninc
проверка уведомлений 2

aurusov commented on warehouse/order_request_model.rao:
```diff
@@ -0,0 +1,56 @@
+
+constant terminateTime = 600 // Время остановки
+constant allowedDeliveryWaitPeriod = 4 // Допустимое время ожидания заказа
+constant defaultDeliveryTime = 3 // Время ожидания в днях
```
Если уж писать размерность, то у всех трех констант.

Commit:
Поправил запрос к базе с учетом нового синтаксиса (aurusov/raoxmodels#16)

Commit:
Поправил модель после добавления dataentity (aurusov/raox-models#16)

Commit:
Описал сущности из базы и успешно прогнал модель (aurusov/raox-models#16)

Commit:
Merge branch 'master' of https://github.com/mikhailmineev/raox-models

# Conflicts:
#	delivery_to_order/warehouse_model.rao

Commit:
Форматирование (aurusov/raox-models#16)

aurusov commented on delivery_to_order/java/domain/Part.java:
```diff
@@ -0,0 +1,49 @@
+	public List<PartRequest> requestedParts;
+
+	@Override
+	public String toString(){
```
Включи автоматическое форматирование при сохранении *.java

aurusov commented on delivery_to_order/order_request_model.rao:
```diff
@@ -0,0 +1,86 @@
+import domain.PartRequest
+import java.util.Iterator
+import domain.Part
+import domain.QOrder
```
Давай сортировать их по имени.

aurusov commented on delivery_to_order/order_request_model.rao:
```diff
@@ -0,0 +1,86 @@
+/*dataprovider corpTerminal = new SqlDataProvider("jdbc:mysql://localhost:3306/corpterminal","observer","compaq")
+	.addEntity("Part")*/
+
+type OrderIterator {
```
Разве мы не так договаривались ?
`OrderIterator` -> `Orders`

aurusov commented on delivery_to_order/warehouse_model.rao:
```diff
@@ -0,0 +1,59 @@
+constant deliveryAmount = 200 // В штуках
+constant itemPrice = 1_000_00 // В копейках
+constant defaultDeliveryTime = 3 // Время ожидания в днях
+constant terminateTime = 600 // Время остановки
```
Нет размерности

aurusov commented on delivery_to_order/warehouse_model.rao:
```diff
@@ -0,0 +1,59 @@
+	int deliveryTime
+}
+
+type WPart {
```
Надо переименовать

aurusov commented on delivery_to_order/warehouse_model.rao:
```diff
@@ -0,0 +1,59 @@
+	int totalPrice
+}
+
+type PartIterator {
```
`PartIterator` -> `Parts`

aurusov commented on delivery_to_order/warehouse_model.rao:
```diff
@@ -0,0 +1,59 @@
+event ItemReceived() {
+	var part = partIterator.iterator.next()
+	var price = calculatePrice(part)
+	WPart.create(part,price)
```
пробел

aurusov commented on delivery_to_order/warehouse_model.rao:
```diff
@@ -0,0 +1,59 @@
+
+event ItemReceived() {
+	var part = partIterator.iterator.next()
+	var price = calculatePrice(part)
```
`var` -> `val` в обоих случаях

aurusov commented on delivery_to_order/warehouse_model.rao:
```diff
@@ -0,0 +1,59 @@
+
+def init() {
+	var query = corpTerminal.getQuery()
+	var part = QPart.part
```
Старайся везде писать `val`

Commit:
Исправил модель не запускается из-за нехватающих сущностей (aurusov/raox-models#16)

Commit:
Исправления по замечаниям (aurusov/raox-models#16)

aurusov commented on delivery_to_order/order_request_model.rao:
```diff
@@ -14,17 +16,21 @@ enum State {REQUESTERD, IN_TRANSIT, ARRIVED}
-	"jdbc:mysql://mikhailmineev.ru:3306/corpterminal?zeroDateTimeBehavior=convertToNull",
-	"jpademo",
+	"jdbc:mysql://mikhailmineev.ru:3306/corpterminal?zeroDateTimeBehavior=convertToNull", 
+	"jpademo", 
```
А это что за пробелы в конце ? Эх, нет у нас форматирования для моделей.

mikhailmineev replied:
Форматирование делает что-то типа
```
dataprovider data = new SqlDataProvider("com.mysql.jdbc.Driver",
	"jdbc:mysql://localhost:3306/corpterminal?zeroDateTimeBehavior=convertToNull", "observer", "compaq", Part, Order,
	PartRequest)
```
Но оно менее читаемо
Собственно, когда я разбивал параметры на строки, тогда и забыл вручную убрать пробелы

mikhailmineev replied:
Решил, как исправить, на данные для подключения создам отдельные константы

aurusov commented:
Посмотрел. Текущее состояние `ok`.

Commit:
Исправление форматирования (aurusov/raox-models#16)

Commit:
Поправил код под добавленный тип cursor (aurusov/raox-models#16)

mikhailmineev commented on delivery_to_order/order_request_model.rao:
```diff
@@ -0,0 +1,85 @@
+//constant password = "5xYB2e6T5Jo7ajA"
+
+dataprovider data = new SqlDataProvider(driver, url, username, password, Part, Order, PartRequest)
+cursor orders = new ArrayList<Part>().iterator()
```
Эта штука по теории должна быть пустой, т.е. `cursor orders`, но тогда при компиляции создается поле `public static Object orders;` Чтобы определился тип, мне приходится давать пустой итератор. Пока не знаю, как исправить это.

aurusov replied:
А как инициализируются ресурсы, к примеру ? Если нужен тип слева, то можно подумать в сторону
`cursor<Part> orders`

Commit:
Подвел код в модели под тот, что на схеме (auusov/raox-models#16)

mikhailmineev commented on delivery_to_order/order_request_model.rao:
```diff
@@ -0,0 +1,160 @@
+		else
+			orders.successes = orders.successes + 1
+		//resetTime
+		//OrderReceived.plan(0) // TODO спросить почему только в эвентах работает 
```
Решил попробовать моделирование со сбросом времени, поскольку вариант с перезапуском модели оказался больно сложным. После утилизации заказа (это значит, что он полностью обработан) хотел сбрасывать время 
и вызывать новое событие. Однако если я планирую событие здесь, то оно не вызывается. Я не могу понять, почему.
Хотя можно логику в событии перенести в функцию и вызывать здесь функцию, но пока не пробовал

aurusov replied:
Сбрасывать время - это костыль. В системе есть бинарная БД, которая хранит состояния модели. Её надо инициализировать с нуля. У неё есть подписчики, например, графики. Они продолжат получать значения ? Еще мы говорили про значения по умолчанию для объектов ? Их тоже надо сбросить в исходное значение. В общем одним сбросом переменной времени не обойтись. Надо выделять в отдельный метод инициализацию модели и дергать его перед итерацией.

aurusov replied:
Перечитал твой вопрос. Кажется, что мне надо объяснить тебе, что такое системное время и как оно продвигается вперед. Голосом.

`OrderReceived.plan(0)` - нельзя запланировать событие в прошлом. Если даже можно, то как ты представляешь его выполнение ?

mikhailmineev replied:
Тогда можно будет завтра обсудить около 8 часов? Сегодня нет возможности

aurusov replied:
Хорошо

mikhailmineev commented on delivery_to_order/order_request_model.rao:
```diff
@@ -0,0 +1,160 @@
+
+def terminateCondition() {
+	return iterationCount >= iterationsTotal && finished || !orders.iterator.hasNext
+	// TODO не пашет скипает на время 6000 и останавливается
```
Я сейчас обрабатываю 30 первых заказов, но обработаться успевает только 10, не совсем понимаю, почему так, где-то на 10-15 днях берет и уходит ко времени 6000, единственная связь тут, раньше была остановка на времени 6000, но теперь её нет.

aurusov replied:
См. трассировку. Можешь расставить логирование внутри событий.

aurusov commented on delivery_to_order/order_request_model.rao:
```diff
@@ -4,82 +4,157 @@ import domain.PartRequest
+}
+
+rule OrderProcessing() {
+	relevant order_ = OrderType.accessible.filter[state == OrderState.REQUESTED].any
```
Если нет конфликта имен, то лучше `order_` -> `order`

Commit:
Правка по вопросам

- вывод процента успешнах заказов и длительности обработки заказов
- если длительность нельзярасчитать, то заказ считается неотгруженым
(aurusov/raox-models#16)

mikhailmineev commented on delivery_to_order/order_request_model.rao:
```diff
@@ -0,0 +1,191 @@
+		val persentage = orders.successes * 100 / (orders.successes + orders.fails+ 0.0)
+		log(
+			"Orders " + orders.successes + ":" + orders.fails + "=" + orders.iterationCount + " Success percentage: " +
+				persentage + "%")
```
Мне нужно выводить % успешных заказов и ~время~ длительность каждого заказа, как это сделать через результаты, я не сообразил, на данный момент пишу в лог, но, возможно, нужно вообще другое решение (например записывать во внешний файл)?

aurusov replied:
Напомни, мы это обсудили голосом на консультации и решили использовать `LogStatistic` или вопрос еще актуален ?

mikhailmineev commented:
Можете провести ревью кода модели?

Commit:
Fix месяц в Calendar начинается с 0 (aurusov/raox-models#16)

Commit:
Исправления по вопросам

- Пустые заказы считаем неуспешными
- Длительность пустых заказов и тех, у которых неуспешные детали,
считаем по длительности до последнего изменения
(aurusov/raox-models#16)

Commit:
Исправления по вопросам

- Чисто пустые заказы игнорятся
Пофиксен учет деталей на складе
(aurusov/raox-models#16)

aurusov commented on delivery_to_order/java/domain/Order.java:
```diff
@@ -0,0 +1,177 @@
+		LocalDate start = getDateOfCreation();
+		LocalDate end = getDateOfRealization();
+		long result = start.until(end, ChronoUnit.DAYS);
+		if(result < 0) {
```
Натрави стандартный форматтер на вс java-исходники.

aurusov commented on delivery_to_order/java/domain/Order.java:
```diff
@@ -0,0 +1,177 @@
+		long result = start.until(end, ChronoUnit.DAYS);
+		if(result < 0) {
+			throw new IllegalStateException("getModificationInterval " + result);
+		}
```
В однострочных if'ах мы не используем скобки.

aurusov commented on delivery_to_order/java/domain/Order.java:
```diff
@@ -0,0 +1,177 @@
+	public List<PartRequest> requestedParts;
+
+	/**
+	 * Дата поступления заказа
```
Название метода говорит за себя лучше комментария. Предлагаю их удалить.

aurusov commented on delivery_to_order/java/domain/Order.java:
```diff
@@ -0,0 +1,177 @@
+	 * @return можно ли вычислить интервал реализации
+	 */
+	public boolean hasRealizationInterval() {
+		return dateOfRealization != null;
```
Кажется странным иметь метод бизнес-логики обработки в классе работы с БД. Предлагаю завести потомка и перенести туда.

Commit:
Исправление форматирования (aurusov/raox-models#16)

Commit:
Загрузка коннектора mysql в проекте (aurusov/raox-models#16)

Commit:
Форматирование (aurusov/raox-models#16)

aurusov commented on delivery_to_order/lib/README.txt:
```diff
@@ -0,0 +1,12 @@
+������ ������ ���������� ���������� mysql-connector-java-5.1.44-bin.jar
+������� � https://dev.mysql.com/downloads/connector/j/
+
+����� �� �������� � ����������� ��-�� ��� ��������
```
Почему не utf-8 ?

Commit:
Encoding (aurusov/raox-models#16)

Commit:
Результаты моделирования в методе finish (aurusov/raox-models#16)

Commit:
Вынес логику из кода сущностей в код модели (aurusov/raox-models#16)

Commit:
Fix refactoring (aurusov/raox-models#16)

Commit:
Переставил импорты, чтобы при ребилде не плыли (aurusov/raox-models#16)

Commit:
Заказы с dateOfDelivery NULL должны игнорироваться
(aurusov/raox-models#16)

Commit:
Рефакторинг (aurusov/raox-models#16)

Commit:
Merge branch 'master' of https://github.com/mikhailmineev/raox-models

# Conflicts:
#	delivery_to_order/order_request_model.rao

Commit:
Добавил простые эксперименты (aurusov/raox-models#16)

Commit:
Модель с запуском экспериментов (aurusov/raox-models#16)

Commit:
Фикс, разные результаты в идентичных прогонах в экспериментх
(aurusov/raox-models#16)

Commit:
Доставка и вычисляемое время ожидания (lekaitow/raox-models#16)

Commit:
Изменения по замечаниям

Добавил баланс
Заказ деталей учитывается в балансе
Эксперимент по коэффициенту количества заказываемых деталей
Расчет длительности ожидания заказа по гистограмме по дням, со
значениями успех (реализован)/неуспех (lekaitow/raox-models#16)

Commit:
Переработал модель, избавился от заказов, забираются детали только для
успешного (lekaitow/raox-models#16)

Commit:
Рефакторинг модели, примерно адекватные значения
(aurusov/raox-models#16)

Commit:
Добавил проекты с моделями as-is и c статистикой основанной на
вероятности заказа конкретной детали (aurusov/raox-models#16)

Commit:
Правка модели для более удобной настройки (aurusov/raox-models#16)

Commit:
Фикс в эксперименте нет влияния наличия детали на складе на его
успешность (aurusov/raox-models#16)
